Fetch system user data on restore

proteusvacuum opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->
The P1 on Monday showed us that having this data stored in duplicate is unnecessary and brittle. 
This changes the way system user data is generated, and ensures that any system keys are maintained out of reach of users. Having this generated in the restore at the time it is needed means the adding and removing location logic is simplified, and there is now only one place to look for system user data keys.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### RISK ASSESSMENT / QA PLAN
<!-- Does this need QA before or after merge? Link QA ticket. -->
Given the high importance of this code in making apps function correctly, I'm going to request QA. 

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

proteusvacuum committed:
Generate location data JIT

proteusvacuum committed:
Dont save locations to user data

proteusvacuum committed:
Lazy migrate user data to delete location keys

proteusvacuum committed:
Save commtrack_supply_point to the user

orangejenny reviewed (approved):
Hooray, I'm very much in favor in this.

I think `commcare_project` could also be added to `user_session_data`, instead of in `create` and `wrap`. What do you think?

proteusvacuum committed:
Get commcare_project JIT

proteusvacuum committed:
Update tests to pull from user_session_data

proteusvacuum committed:
Only add session data when needed

millerdev reviewed (approved):
Oooh, this looks like a very good change!

proteusvacuum committed:
Send back system userdata keys in API

proteusvacuum committed:
Add project for CommCareUser only

proteusvacuum committed:
Add project for users with any domain membership

proteusvacuum committed:
User data in repeaters

kaapstorm committed:
Nit: Using `get()` isn't safer here

Using `get()` here doesn't avoid an error. If the key 'user_data' is not
found, this would just cause an AttributeError 3 or 7 lines later.

I chose `['user_data']` over `get('user_data', {})` because this code is
called a lot and works as it is. If 'user_data' were to be missing, this
code ought to throw an error (and the error should be a KeyError, not an
AttributeError).

kaapstorm commented:
I was going to add a comment, and then figured it would be easier just to add a commit. I hope the commit description is self-explanatory.

orangejenny committed:
Merge branch 'master' into fr/user-data-jit

proteusvacuum committed:
Remove fields not present in model when wrapping

proteusvacuum commented on corehq/apps/users/models.py:
```diff
@@ -1073,6 +1073,14 @@ def wrap(cls, data, should_save=False):
             should_save = True
 
+        # Wrapping a doc from ES adds these extra fields which shouldn't be saved
+        if "__group_ids" in data:
```
@sravfeyn FYI https://dimagi-dev.atlassian.net/browse/QA-1835

sravfeyn replied:
```suggestion
        for attr in ["__group_ids", "__group_names", "user_data_es"]:
                if attr in data:
                        del data[attr]
```

proteusvacuum replied:
Hmm... I actually think saving docs that are coming in from ES is dangerous, since we could be saving an old version. I'm going to think about this a bit more.

proteusvacuum committed:
Dont save docs coming from ES

proteusvacuum committed:
Fix typo

proteusvacuum committed:
Delete users after tests

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/user-data-jit

proteusvacuum committed:
Include profiles

proteusvacuum committed:
Include commcare_project only if there is a domain

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/user-data-jit

proteusvacuum committed:
Remove commcare-project from metadata tests

proteusvacuum committed:
More commcare_projects

proteusvacuum committed:
Merge branch 'master' into fr/user-data-jit

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/user-data-jit

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2117,18 +2152,10 @@ def unset_location(self, fall_back_to_next=False, commit=True):
             self.set_location(SQLLocation.objects.get(location_id=new_primary_location_id))
         else:
-            self.pop_metadata('commcare_location_id', None)
-            self.pop_metadata('commtrack-supply-point', None)
```
I think this needs to be replaced with
```python
self.commtrack_supply_point = None
```

proteusvacuum replied:
good catch, thanks

esoergel reviewed (approved):
Nice change - this stuff is super messy

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/user-data-jit

proteusvacuum commented on corehq/apps/users/models.py:
```diff
@@ -1214,6 +1241,29 @@ def user_session_data(self):
+        except AttributeError:
+            domains = self.get_domains()
+            if domains:
+                domain = domains[0]
```
Yeah, I didn't want to change behaviour here which is why I did this. I'm not sure if you can really use user_data on a web user...

orangejenny replied:
I dug into this question as part of the user profiles work. It looks like web users do not support user data, but they did support it in the past, at least for some certain projects, which is why `user_data` is an attribute of `CouchUser` and not `CommCareUser`. I want to say ews-ghana had some kind of extended web user that could have user data. I looked at constraining user data to mobile users and ended up deciding it would increase my scope too much.

proteusvacuum commented on corehq/apps/users/models.py:
```diff
@@ -1050,6 +1052,31 @@ def wrap(cls, data, should_save=False):
+
+        # Wrapping a doc that comes out of ES shouldn't be saved here, even if deprecated keys are present
+        if any(es_param in data for es_param in ["__group_ids", "__group_names", "user_data_es"]):
+            should_save = False
```
do you mean instead of doing this lazily, I should do it in a migration? I think that's probably preferable now, good idea.

esoergel replied:
Yeah, or doing it lazily up front for an easier roll-out, _then_ doing the migration.  That can be useful if the migration might take a while and isn't backwards compatible, so you don't get errors in the transition.  I haven't really thought through which would be better here, was just flagging that it's of course preferable to not do permanent lazy migrations.

orangejenny replied:
I'm in favor of doing real migrations, just want to make sure that if you're considering a real migration, you're accounting for the number of users on CAS.

proteusvacuum replied:
Yeah, I'm also a little worried that this lazy migration is also going to add a ton of writes to couch right after this change is merged (one for every login, although maybe that isn't an issue?). 

How have we handled user migrations on CAS other times? 

Neither migration is totally necessary for this feature to work (as any stale values will be ignored when the restore is generated), but we would end up with a very confusing set of data if we don't migrate.

orangejenny replied:
I don't know if we've ever run a script across all couch users on CAS. @sravfeyn @snopoke we're talking about doing a migration that would remove a couple of user data keys from couch users. Do you have any wisdom to share?

snopoke replied:
If the lazy migration is not necessary then I'd rather have it removed so that we can control when the user updates happen. On CAS this is particularly important since updating every user on their first request after this is merged will likely bring CouchDB to it's knees.

Other than that I don't see an issue with doing an off-deploy migration. It may take a while on ICDS but we can do it off peak.

proteusvacuum committed:
reset commtrack supply point when unsetting location

proteusvacuum commented on corehq/apps/users/models.py:
```diff
@@ -1214,6 +1241,29 @@ def user_session_data(self):
+        if self.assigned_location_ids:
+            session_data['{}_location_ids'.format(SYSTEM_PREFIX)] = user_location_data(self.assigned_location_ids)
+        if self.commtrack_supply_point:
+            session_data['commtrack-supply-point'] = self.commtrack_supply_point
```
Do you mean this is an incorrect change (i.e. it shouldn't be applied to web users), or is just confusing to read because it only applies to mobile workers?

esoergel replied:
Sorry, yeah, that was unclear.  I mean that I think this is a behavior change.  That is, in the current state of the code, the location-related fields are all added to the metadata only for mobile workers.  For example, `WebUser.set_location` doesn't modify `self.metadata`, but `CommCareUser.set_location` does.

This shouldn't actually matter, because `WebUser` locations are set on the domain membership, but due to some accident of history, `location_id` and `assigned_location_ids` are set on `CouchUser` instead of `CommCareUser`, so there's probably some data out there with those fields set also for WebUsers.

Regarding correctness, I think putting any domain-specific data in `WebUser.user_session_data` (which is _not_ domain specific) is incorrect, as per the comment thread above, but I do agree it's safer for this PR's scope to keep that behavior consistent, as you've done with `commcare_project`.

proteusvacuum committed:
Add note about web user domains in session data

snopoke commented on corehq/apps/users/models.py:
```diff
@@ -1041,6 +1041,8 @@ class CouchUser(Document, DjangoUserMixin, IsMemberOfMixin, EulaMixin):
     login_attempts = IntegerProperty(default=0)
     attempt_date = DateProperty()
 
+    commtrack_supply_point = StringProperty()
```
based on the discussion below, should this be on CouchUser or only on CommCareUser?

snopoke reviewed (commented):
bump?

proteusvacuum commented:
@snopoke ... we never released this because of the effect it might have had on CAS... but maybe we should revisit? I had built this originally while working on what became USH... maybe USH wants to pick this up now?

orangejenny commented:
I have some vague memories of this. I had thought it ran into issues in QA. @proteusvacuum do you remember anymore about what the potential issues with CAS were?

proteusvacuum commented:
@orangejenny I think the issue was the lazy migration and the strain it was going to put on couch: https://github.com/dimagi/commcare-hq/pull/28456#discussion_r501983511 
I think I lost steam after that point as it sounded like it was going to be a challenge to run this for CAS
